src/proxy_handler.cpp

```diff
@@ -0,0 +1,206 @@
+    }
+
+    // Read the response headers, which are terminated by a blank line.
+    boost::asio::read_until(socket, response_, "\r\n\r\n");
+
+    std::string header;
+    size_t index;
+    size_t cr;
+    std::string new_host = "";
+    std::vector<std::pair<std::string, std::string>> header_vector;
+    while (std::getline(response_stream, header) && header != "\r") {
+      index = header.find(':', 0);
+      cr = header.find('\r', index);
+      if (index != std::string::npos) {
+        if (header.substr(0, index).compare("Location") || status_code != 302) {
+        header_vector.push_back(std::make_pair(header.substr(0, index), header.substr(index+2, cr - (index + 2))));
+        }
+        else {
+          new_host = header.substr(index+2, cr - (index + 2));
+          
+          if (!strncmp(new_host.c_str(), "http://", 7)) {
+            new_host = new_host.substr(7);
+          } else if (!strncmp(new_host.c_str(), "https://", 8)) {
+            new_host = new_host.substr(8);
+          }
+          size_t first_slash = new_host.find("/");
+          if (first_slash != std::string::npos) {
+            new_host = new_host.substr(0, first_slash);
+            redirect_uri_ = new_host.substr(first_slash);
+          }
+          
+          proxy_host_ = new_host;
+          
+        }
+      }  
+    }
+    if (!new_host.empty()) {
+      return HandleRequestHelper(request, response, (count + 1));
+    }
+    else {
+      for (unsigned i = 0; i < header_vector.size(); i++) {
+        response->AddHeader(header_vector[i].first, header_vector[i].second);
+      }
+    }
+    std::cout << "\n";
+
+    // Write whatever content we already have to output.
+    if (response_.size() > 0) {
+      std::ostringstream ss;
+      ss << &response_;
+      body_ += ss.str();
+    }
+    
+    // Read until EOF, writing data to output as we go.
+    boost::system::error_code error;
+    while (boost::asio::read(socket, response_, boost::asio::transfer_at_least(1), error)) {
+      std::ostringstream ss;
+      ss << &response_;
+      body_ += ss.str();
+    }
+
+    if (error != boost::asio::error::eof)
+      throw boost::system::system_error(error);
+    if (uri_prefix_.size() > 1) { 
+      boost::regex r("src\\s*=\\s*\"((?!http)[^\"]*)\"");
+      std::string fmt = "src=\"" + uri_prefix_ + "\\1\"";
+      boost::regex r2("href\\s*=\\s*\"((?!http)[^\"]*)\"");
+      std::string fmt2 = "href=\"" + uri_prefix_ + "\\1\"";
+      std::string replaced  = boost::regex_replace(body_, r, fmt, boost::match_default | boost::format_sed);
+      body_ = boost::regex_replace(replaced, r2, fmt2, boost::match_default | boost::format_sed);
+    }
+
+    response->SetBody(body_);
+    std::string length;
+    std::ostringstream temp;
+    temp  <<  ((int) body_.size());
+    length = temp.str();
+    response->AddHeader("Content-Length", length);
+  } catch (std::exception& e) {
+    std::cout << "Exception: " << e.what() << "\n";
```

Comment:
If there's an exception, we should do more than print the error. Return a non-ok RequestHandler status code.